File: 202.HappyNumber/Happy.java
public class Solution {
    public boolean isHappy(int n) {
        Set<Integer> nums = new HashSet<>();

        while (n > 1) {
            if (nums.contains(n)) {
                return false;
            }
            nums.add(n);

            int tmp = n;
            int sum = 0;
            while (tmp > 0) {
                int d = tmp % 10;
                tmp /= 10;
                sum += d * d;
            }

            n = sum;
        }

        return true;
    }
}

File: 202.HappyNumber/happy.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """

        nums = {}
        while n > 1:
            if n in nums:
                return False
            nums[n] = 0
            tmp = n
            s = 0
            while tmp > 0:
                d = tmp % 10
                tmp //= 10
                s += d * d
            n = s

        return True

if __name__ == '__main__':
    n = 20
    print(Solution().isHappy(n))
